web/ui/react-app/src/Graph.tsx

```diff
@@ -238,64 +236,78 @@ class Graph extends PureComponent<GraphProps, GraphState> {
     if (!this.chartRef.current) {
       return;
     }
-    const selectedData = this.getData()[this.state.selectedSeriesIndex!];
     this.destroyPlot();
-    $.plot($(this.chartRef.current), selectedData ? [selectedData] : this.getData(), this.getOptions());
+    this.$chart = $.plot($(this.chartRef.current), this.state.chartData, this.getOptions());
   };
 
   destroyPlot() {
-    const chart = $(this.chartRef.current!).data('plot');
-    if (chart !== undefined) {
-      chart.destroy();
+    const $chart = $(this.chartRef.current!).data('plot');
+    if ($chart !== undefined) {
+      $chart.destroy();
     }
   }
 
-  handleSeriesSelect = (index: number) => () => {
-    const { selectedSeriesIndex } = this.state;
-    this.setState({ selectedSeriesIndex: selectedSeriesIndex !== index ? index : null });
+  plotSetAndDraw = (data: GraphSeries[] = this.state.chartData) => {
+    if (this.rafID) {
+      cancelAnimationFrame(this.rafID);
+    }
+    this.rafID = requestAnimationFrame(() => {
+      this.$chart.setData(data);
+      this.$chart.draw();
+    });
   };
 
-  handleSeriesHover = (index: number) => () => {
-    this.setState({ hoveredSeriesIndex: index });
+  handleSeriesSelect = (index: number) => () => {
+    this.setState(({ selectedSeriesIndex, chartData }) => {
+      this.plotSetAndDraw(
```

Comment:
Is it "normal" to put side effects like this into the `setState()` updater function? Just for my understanding, you probably had a good reason to call `this.plotSetAndDraw(...)` directly in `setState()` rather than outside of it?

Reply:
Yes, so this is same as
```
const { selectedSeriesIndex, chartData } = this.state;
this.plotSetAndDraw(
      selectedSeriesIndex === index
          ? chartData.map(this.toHoverColor(index))
           : chartData.slice(index, index + 1)
);
this.setState(
  { selectedSeriesIndex: selectedSeriesIndex === index ? null : index }
)
```
I just decided to use the callback version of setState which receives the prev state values and since ```plotSetAndDraw``` is not depending (in this case) on the next state is absolutely fine to be triggered from there. Thus looks more like a single unit.
Do you have any concerns about this?

Reply:
When I approach this kind of question I always think mostly about whoever reads this code and what they would expect, and what would surprise them.

In this case, it's surprising for a reader to have extra side effects in a `setState()` call, which is (to my knowledge) normally only used to do direct state transformations on `this.state`, but nothing else besides that. So I'd prefer separating the two concerns and doing the plotting outside of `setState()`. Never be afraid of a tiny bit of extra code if it's less surprising :)

Reply:
Yes, I agree. I was thinking from another point of view but your suggestion makes more sense.